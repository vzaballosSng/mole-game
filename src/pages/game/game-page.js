import { LitElement, html } from 'lit';
import { gamePageStyles } from './game-page-styles.js';
import { commonStyle } from '../../styles/common.js';
import { MoleManager } from '../../managers/mole-manager/mole-manager.js'
import {
  TIME_LEVEL,
  TILES_NUMBER,
  STATES,
  MOLE_STATUS,
  LEVELS_LIST,
} from '../../utils/constants.js';
import '../../components/custom-header/custom-header.js';
import '../../components/mole-tile/mole-tile.js';
import '../../components/custom-button/custom-button.js';
import '../../components/notification-message/notification-message.js';

export class GamePage extends LitElement {
  static get styles() {
    return [gamePageStyles, commonStyle];
  }

  /**
   * @property {String} name - Player's name
   */
  static get properties() {
    return {
      playerName: { type: String, attribute: 'player-name' },
      _points: { type: Number },
      _buttonTitle: { type: String },
      _tilesAltArray: { type: Array },
      _levelList: { type: Array },
      _levelListSelector: { type: Array },
      _moleRender: { type: Object },
      _openNotification: {type: Boolean }
    };
  }

  constructor() {
    super();
    this.playerName = '';
    this._moleManager = new MoleManager();
    this._levelSelected = 'Low';
    this._levelList = this._moleManager.getLevels();
    this._levelListSelector = [...LEVELS_LIST];
    this._timerId = null;
    this._buttonTitle = STATES.play;
    this._points = 0;
    this._moleTilesList = '';
    this._previousTileSelected = false;
    this._currentTileSelected = 0;
    this._tilesAltArray = this._moleManager.generateTiles(TILES_NUMBER);
    this._listAux = this._moleManager.generateTiles(TILES_NUMBER);
    this._moleRender = {
      list: [...this._tilesAltArray],
    };
    this._gameState = STATES.stop;
    this._openNotification = false;
  }

  _handleButton() {
    if (this._buttonTitle === STATES.play) {
      this._buttonTitle = STATES.stop;
      this._startGame();
    } else {
      this._buttonTitle = STATES.play;
      this._gameState = STATES.stop;
      this._pauseGame();
      this._points = this._moleManager.restartPoints();
      this._resetTiles();
    }
  }

  _pauseGame() {
    if (this._timerId) {
      clearInterval(this._timerId);
      this._timerId = null;
    }
    return true;
  }

  _startGame() {
    this._timerId =  setInterval(() => this._showMoleAlt(), TIME_LEVEL[this._levelSelected]);
    this._gameState = STATES.play;
  }

  _showNotification() {
    this._openNotification = true;
    setTimeout(() => {
      this._openNotification = false;
    }, 1500);
  }

  _resetTiles() {
    console.log(this._listAux);
    this._moleRender = {
      ...this._moleRender,
      list: [...this._listAux]
    };
  }

  _showMoleAlt() {
    this._currentTileSelected = this._moleManager.getRandomTile(TILES_NUMBER, this._previousTileSelected);
    if (this._previousTileSelected) {
      this._tilesAltArray[this._previousTileSelected].show = false;
    }
    this._previousTileSelected = this._currentTileSelected;
    this._tilesAltArray[this._currentTileSelected].show = true;
    this._tilesAltArray[this._currentTileSelected].status = MOLE_STATUS.alive;

    this._moleRender = {
      ...this._moleRender,
      list: [...this._tilesAltArray]
    };
  }

  _clickMole(ev) {
    if (ev.target.showMole && this._tilesAltArray[this._currentTileSelected].status === MOLE_STATUS.alive) {
      this._savePoints();
      this._tilesAltArray[this._currentTileSelected].status = MOLE_STATUS.dead;
      this._showNotification();
      // navigator.vibrate(200);
    }
  }

  _savePoints() {
    this._points += this._moleManager.getPointsByLevel(this._levelSelected)
  }

  _setLevel(ev) {
    this._levelSelected = ev.detail.value;

    if (this._gameState === STATES.play) {
      this._pauseGame();
      this._startGame();
    }
  }

  render() {
    return html`
      <custom-header
        @change-select-value=${this._setLevel}
        player-name="${this.playerName}"
      >
        <div slot="left-side">
          <span class="player-name">${this.playerName}</span>
        </div>
        <div slot="right-side">
          <custom-select
            label-name="Level:"
            .listOptions=${this._levelListSelector}
          ></custom-select>
        </div>
      </custom-header>
      <p class="points">Points: ${this._points}</p>
      <section class="mole-container">
        ${this._moleRender.list.map(item => {
          return html` 
            <mole-tile 
              id=${item.id}
              ?show-mole=${item.show}
              @click=${this._clickMole}
            ></mole-tile>`;
        })}
      </section>
      <div class="button-container">
        <custom-button 
          data-state=${this._state}
          @click=${this._handleButton}
        >${this._buttonTitle}</custom-button>
      </div>
      
      <notification-message message="Well done" ?show=${this._openNotification}></notification-message>
    `;
  }
}

customElements.define('game-page', GamePage);
