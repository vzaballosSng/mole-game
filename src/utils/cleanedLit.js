// eslint-disable-next-line no-undef
const { prettyDOM } = require('@testing-library/dom');

export const cleanedLit = (htmlElement) => {
	const domHTML = prettyDOM(htmlElement, Infinity);
	const cleaned = domHTML.replace(/"<!--\?lit\$.*?-->"/g, '""');
	return cleaned;
};
export const cleanedLitElements = (htmlElements) => {
	const cleaned = htmlElements.map((htmlElement) => {
		return cleanedLit(htmlElement);
	});
	return cleaned;
};
