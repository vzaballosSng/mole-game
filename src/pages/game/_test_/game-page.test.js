
import { html, fixture } from '@open-wc/testing';
import { expect } from '@jest/globals';
import { cleanedLitElements } from '../../../utils/cleanedLit.js';
import { getAllElementsAndShadowRoots } from 'shadow-dom-testing-library';
import '../game-page.js';

describe('game-page', () => {
	it('game-page matches the snapshot', async () => {
		const el = await fixture(
			html`<game-page player-name"test"></game-page>`
		);
		const data = getAllElementsAndShadowRoots(el, '*');
		const dataCleaned = cleanedLitElements(data);
		expect(dataCleaned).toMatchSnapshot();
	});
});
