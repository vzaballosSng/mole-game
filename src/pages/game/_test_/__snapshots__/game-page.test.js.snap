// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`game-page game-page matches the snapshot 1`] = `
[
  "[36m<game-page[39m
  [33mplayer-name"test"[39m=[32m""[39m
[36m/>[39m",
  "ShadowRoot {}",
  "ShadowRoot {}",
  "[36m<header[39m
  [33mclass[39m=[32m"header"[39m
[36m>[39m
  [0m
      [0m
  [36m<slot[39m
    [33mname[39m=[32m"left-side"[39m
  [36m/>[39m
  [0m
      [0m
  [36m<slot[39m
    [33mname[39m=[32m"right-side"[39m
  [36m/>[39m
  [0m
		[0m
[36m</header>[39m",
  "[36m<slot[39m
  [33mname[39m=[32m"left-side"[39m
[36m/>[39m",
  "[36m<slot[39m
  [33mname[39m=[32m"right-side"[39m
[36m/>[39m",
  "[36m<style>[39m
  [0m
	:host {
    display: block;
    box-sizing: border-box;
    background-color: #a0522d;
    padding: 1.25rem;
    width: 100%;
	}

  .header {
    display: flex;
    justify-content: space-between;
    padding: 0;
  }
[0m
[36m</style>[39m",
  "[36m<div[39m
  [33mslot[39m=[32m"left-side"[39m
[36m>[39m
  [0m
          [0m
  [36m<span[39m
    [33mclass[39m=[32m"player-name"[39m
  [36m/>[39m
  [0m
        [0m
[36m</div>[39m",
  "[36m<span[39m
  [33mclass[39m=[32m"player-name"[39m
[36m/>[39m",
  "[36m<div[39m
  [33mslot[39m=[32m"right-side"[39m
[36m>[39m
  [0m
          [0m
  [36m<custom-select[39m
    [33mlabel-name[39m=[32m"Level:"[39m
  [36m/>[39m
  [0m
        [0m
[36m</div>[39m",
  "ShadowRoot {}",
  "[36m<label[39m
  [33mfor[39m=[32m"level"[39m
[36m>[39m
  [0mLevel:[0m
[36m</label>[39m",
  "[36m<select[39m
  [33maria-label[39m=[32m"level"[39m
  [33maria-labelledby[39m=[32m"level"[39m
  [33mid[39m=[32m"level"[39m
  [33mname[39m=[32m"select-level"[39m
[36m>[39m
  [0m
        [0m
  [0m [0m
  [36m<option[39m
    [33mvalue[39m=[32m"Low"[39m
  [36m>[39m
    [0mLow[0m
  [36m</option>[39m
  [0m [0m
  [36m<option[39m
    [33mvalue[39m=[32m"Medium"[39m
  [36m>[39m
    [0mMedium[0m
  [36m</option>[39m
  [0m [0m
  [36m<option[39m
    [33mvalue[39m=[32m"High"[39m
  [36m>[39m
    [0mHigh[0m
  [36m</option>[39m
  [0m
			[0m
[36m</select>[39m",
  "[36m<option[39m
  [33mvalue[39m=[32m"Low"[39m
[36m>[39m
  [0mLow[0m
[36m</option>[39m",
  "[36m<option[39m
  [33mvalue[39m=[32m"Medium"[39m
[36m>[39m
  [0mMedium[0m
[36m</option>[39m",
  "[36m<option[39m
  [33mvalue[39m=[32m"High"[39m
[36m>[39m
  [0mHigh[0m
[36m</option>[39m",
  "[36m<style>[39m
  [0m
  label {
    margin-right: .9375rem;
  }

  select {
    padding: .625rem 0;
    border: none;
    border-bottom: 2px solid white;
    color: white;
    font-size: 1rem;
    outline: none;
    background: transparent;
  } 
[0m
[36m</style>[39m",
  "[36m<p[39m
  [33mclass[39m=[32m"points"[39m
[36m>[39m
  [0mPoints: [0m
  [0m0[0m
[36m</p>[39m",
  "[36m<section[39m
  [33mclass[39m=[32m"mole-container"[39m
[36m>[39m
  [0m
        [0m
  [0m 
            [0m
  [36m<mole-tile[39m
    [33mid[39m=[32m"mole0"[39m
  [36m/>[39m
  [0m 
            [0m
  [36m<mole-tile[39m
    [33mid[39m=[32m"mole1"[39m
  [36m/>[39m
  [0m 
            [0m
  [36m<mole-tile[39m
    [33mid[39m=[32m"mole2"[39m
  [36m/>[39m
  [0m 
            [0m
  [36m<mole-tile[39m
    [33mid[39m=[32m"mole3"[39m
  [36m/>[39m
  [0m 
            [0m
  [36m<mole-tile[39m
    [33mid[39m=[32m"mole4"[39m
  [36m/>[39m
  [0m 
            [0m
  [36m<mole-tile[39m
    [33mid[39m=[32m"mole5"[39m
  [36m/>[39m
  [0m 
            [0m
  [36m<mole-tile[39m
    [33mid[39m=[32m"mole6"[39m
  [36m/>[39m
  [0m 
            [0m
  [36m<mole-tile[39m
    [33mid[39m=[32m"mole7"[39m
  [36m/>[39m
  [0m 
            [0m
  [36m<mole-tile[39m
    [33mid[39m=[32m"mole8"[39m
  [36m/>[39m
  [0m
      [0m
[36m</section>[39m",
  "ShadowRoot {}",
  "[36m<div[39m
  [33mclass[39m=[32m"tile"[39m
[36m>[39m
  [0m
      [0m
  [0m
    [0m
[36m</div>[39m",
  "[36m<style>[39m
  [0m
	.tile {
		width: 6.25rem;
		height: 6.25rem;
		border: 1px solid black;
    background-color: #000000c9;
	}

	.bg-mole {
    width: 6.25rem;
		height: 6.25rem;
		background-image: url('public/molebg.png');
    background-repeat: no-repeat;
    background-size: cover;
    background-position: center;
	}

  .player-name {
    color: white;
    line-height: 2rem;
  }

  .tile p {
    color: black;
  }
[0m
[36m</style>[39m",
  "ShadowRoot {}",
  "[36m<div[39m
  [33mclass[39m=[32m"tile"[39m
[36m>[39m
  [0m
      [0m
  [0m
    [0m
[36m</div>[39m",
  "[36m<style>[39m
  [0m
	.tile {
		width: 6.25rem;
		height: 6.25rem;
		border: 1px solid black;
    background-color: #000000c9;
	}

	.bg-mole {
    width: 6.25rem;
		height: 6.25rem;
		background-image: url('public/molebg.png');
    background-repeat: no-repeat;
    background-size: cover;
    background-position: center;
	}

  .player-name {
    color: white;
    line-height: 2rem;
  }

  .tile p {
    color: black;
  }
[0m
[36m</style>[39m",
  "ShadowRoot {}",
  "[36m<div[39m
  [33mclass[39m=[32m"tile"[39m
[36m>[39m
  [0m
      [0m
  [0m
    [0m
[36m</div>[39m",
  "[36m<style>[39m
  [0m
	.tile {
		width: 6.25rem;
		height: 6.25rem;
		border: 1px solid black;
    background-color: #000000c9;
	}

	.bg-mole {
    width: 6.25rem;
		height: 6.25rem;
		background-image: url('public/molebg.png');
    background-repeat: no-repeat;
    background-size: cover;
    background-position: center;
	}

  .player-name {
    color: white;
    line-height: 2rem;
  }

  .tile p {
    color: black;
  }
[0m
[36m</style>[39m",
  "ShadowRoot {}",
  "[36m<div[39m
  [33mclass[39m=[32m"tile"[39m
[36m>[39m
  [0m
      [0m
  [0m
    [0m
[36m</div>[39m",
  "[36m<style>[39m
  [0m
	.tile {
		width: 6.25rem;
		height: 6.25rem;
		border: 1px solid black;
    background-color: #000000c9;
	}

	.bg-mole {
    width: 6.25rem;
		height: 6.25rem;
		background-image: url('public/molebg.png');
    background-repeat: no-repeat;
    background-size: cover;
    background-position: center;
	}

  .player-name {
    color: white;
    line-height: 2rem;
  }

  .tile p {
    color: black;
  }
[0m
[36m</style>[39m",
  "ShadowRoot {}",
  "[36m<div[39m
  [33mclass[39m=[32m"tile"[39m
[36m>[39m
  [0m
      [0m
  [0m
    [0m
[36m</div>[39m",
  "[36m<style>[39m
  [0m
	.tile {
		width: 6.25rem;
		height: 6.25rem;
		border: 1px solid black;
    background-color: #000000c9;
	}

	.bg-mole {
    width: 6.25rem;
		height: 6.25rem;
		background-image: url('public/molebg.png');
    background-repeat: no-repeat;
    background-size: cover;
    background-position: center;
	}

  .player-name {
    color: white;
    line-height: 2rem;
  }

  .tile p {
    color: black;
  }
[0m
[36m</style>[39m",
  "ShadowRoot {}",
  "[36m<div[39m
  [33mclass[39m=[32m"tile"[39m
[36m>[39m
  [0m
      [0m
  [0m
    [0m
[36m</div>[39m",
  "[36m<style>[39m
  [0m
	.tile {
		width: 6.25rem;
		height: 6.25rem;
		border: 1px solid black;
    background-color: #000000c9;
	}

	.bg-mole {
    width: 6.25rem;
		height: 6.25rem;
		background-image: url('public/molebg.png');
    background-repeat: no-repeat;
    background-size: cover;
    background-position: center;
	}

  .player-name {
    color: white;
    line-height: 2rem;
  }

  .tile p {
    color: black;
  }
[0m
[36m</style>[39m",
  "ShadowRoot {}",
  "[36m<div[39m
  [33mclass[39m=[32m"tile"[39m
[36m>[39m
  [0m
      [0m
  [0m
    [0m
[36m</div>[39m",
  "[36m<style>[39m
  [0m
	.tile {
		width: 6.25rem;
		height: 6.25rem;
		border: 1px solid black;
    background-color: #000000c9;
	}

	.bg-mole {
    width: 6.25rem;
		height: 6.25rem;
		background-image: url('public/molebg.png');
    background-repeat: no-repeat;
    background-size: cover;
    background-position: center;
	}

  .player-name {
    color: white;
    line-height: 2rem;
  }

  .tile p {
    color: black;
  }
[0m
[36m</style>[39m",
  "ShadowRoot {}",
  "[36m<div[39m
  [33mclass[39m=[32m"tile"[39m
[36m>[39m
  [0m
      [0m
  [0m
    [0m
[36m</div>[39m",
  "[36m<style>[39m
  [0m
	.tile {
		width: 6.25rem;
		height: 6.25rem;
		border: 1px solid black;
    background-color: #000000c9;
	}

	.bg-mole {
    width: 6.25rem;
		height: 6.25rem;
		background-image: url('public/molebg.png');
    background-repeat: no-repeat;
    background-size: cover;
    background-position: center;
	}

  .player-name {
    color: white;
    line-height: 2rem;
  }

  .tile p {
    color: black;
  }
[0m
[36m</style>[39m",
  "ShadowRoot {}",
  "[36m<div[39m
  [33mclass[39m=[32m"tile"[39m
[36m>[39m
  [0m
      [0m
  [0m
    [0m
[36m</div>[39m",
  "[36m<style>[39m
  [0m
	.tile {
		width: 6.25rem;
		height: 6.25rem;
		border: 1px solid black;
    background-color: #000000c9;
	}

	.bg-mole {
    width: 6.25rem;
		height: 6.25rem;
		background-image: url('public/molebg.png');
    background-repeat: no-repeat;
    background-size: cover;
    background-position: center;
	}

  .player-name {
    color: white;
    line-height: 2rem;
  }

  .tile p {
    color: black;
  }
[0m
[36m</style>[39m",
  "[36m<div[39m
  [33mclass[39m=[32m"button-container"[39m
[36m>[39m
  [0m
        [0m
  [36m<custom-button[39m
    [33mdata-state[39m=[32m""[39m
  [36m>[39m
    [0mPLAY[0m
  [36m</custom-button>[39m
  [0m
      [0m
[36m</div>[39m",
  "ShadowRoot {}",
  "[36m<button[39m
  [33mclass[39m=[32m"short"[39m
[36m>[39m
  [0m
			  [0m
  [36m<slot />[39m
  [0m
			[0m
[36m</button>[39m",
  "[36m<slot />[39m",
  "[36m<style>[39m
  [0m
	button {
		background-color: #a0522d;
		border: none;
    border-radius: 10px;
		color: white;
    cursor: pointer;
		padding: 15px 10px;
		text-align: center;
		text-decoration: none;
		font-size: 1rem;
		min-width: 6.25rem;
		height: 3.125rem
	}
[0m
[36m</style>[39m",
  "ShadowRoot {}",
  "[36m<p[39m
  [33mclass[39m=[32m"message"[39m
[36m>[39m
  [0mWell done[0m
[36m</p>[39m",
  "[36m<style>[39m
  [0m
	:host {
    background-color: #050605a3;
    bottom: 0;
    color: white;

		height: 1.875rem;
    opacity: 0;
    padding: 1.25rem;
    position: fixed;
    width: 100%;
	}

  :host([show]) {
    opacity: 1;
    transition: opacity 1s ease;
  }

  .message {
    margin: 0;
    line-height: 2rem;
  }
[0m
[36m</style>[39m",
  "[36m<style>[39m
  [0m
  :host {
    display: block;
    background: linear-gradient(to bottom, #7ec850, #4caf50, #388e3c);
    height: 100vh;
  }

  .button-container {
    margin-top: 1rem;
    text-align: center;
  }

	.mole-container {
    display: grid;
    grid-template-columns: repeat(3, 100px);
    grid-template-rows: repeat(3, 100px);
    justify-content: center;
    align-items: center;
    gap: 1.875rem;
  }

	.points {
		color: white;
    font-size: 1.5rem;
    text-align: right;
    padding-right: 1.25rem;
	}

[0m
[36m</style>[39m",
  "[36m<style>[39m
  [0m

	h1 {
		font-size: 2rem;
	}
[0m
[36m</style>[39m",
]
`;
